### PR TITLE
adds rename-search-fields fn

### PR DESCRIPTION
**Pull request by agzam**
Automatic remapping `source` to `source.text` in queries

> **Epic** https://github.com/advthreat/iroh/issues/5929
> Close https://github.com/advthreat/iroh/issues/5999
> Related https://github.com/advthreat/iroh/issues/5938

<a name="qa">[§](#qa)</a> QA
============================
No QA is needed. Will be covered within https://github.com/advthreat/iroh/issues/5929

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

File: src/ctia/stores/es/crud.clj [hunk 1]
```diff
@@ -453,6 +453,18 @@ It returns the documents with full hits meta data including the real index in wh
...
                                            get-in-config))))))
+(s/defn rename-search-fields :- (s/maybe {s/Keyword [s/Str]})
+  "Automatically translates keyword fields to use underlying text field.
+
+   ES doesn't like when different types of tokens get used in the same query. To deal with
+   that, we create a nested field of type 'text', see:
+   `ctia.stores.es.mapping/searchable-token`. This should be opaque - caller shouldn't
+   have to explicitly instruct API to direct query to the nested field."
+  [fields :- (s/maybe [s/Any])]
+  (let [search-fields-mapping {"source" "source.text"}]
```
Comment (ereteog) [inline src/ctia/stores/es/crud.clj:464]: you need to include all fields which are in default search and mapped as searchable-token: https://github.com/threatgrid/ctia/pull/1155/files

**Review by ereteog — COMMENTED** [review]
We need to change the mappings of token default fields to searchable-token and rename them all. There is at least `id` and `observable.value` in the [common mapping definition](https://github.com/threatgrid/ctia/blob/master/src/ctia/stores/es/mapping.clj).

File: src/ctia/stores/es/query.clj [hunk 2]
```diff
@@ -53,3 +55,81 @@
...
+           res))
+       (if (map? val)
+         (let [path (if (not= :properties key)
+                      (conj prev (name key))
+                      prev)]
+           (searchable-fields-map-impl path val res))
+         res)))
+   result m))
+
+(s/defn searchable-fields-map :- {s/Str s/Str}
```
Comment (DeLaGuardo) [inline src/ctia/stores/es/query.clj:87]: Do we have a test for that function somewhere? It is a bit hard to review the function without some example of usage 😢

Comment (ereteog) [reply src/ctia/stores/es/query.clj:87]: In addition, it's sensitive enough to be unit tested and check the edge cases.

**Review by ereteog — COMMENTED** [review]
Do not forget to define searchable-token mappings

> We need to change the mappings of token default fields to searchable-token and rename them all. There is at least id and observable.value in the common mapping definition.

We need unit tests on the fields rewrite to gain confidence on it (now I cannot tell if it will works with edge cases)

File: src/ctia/stores/es/query.clj [hunk 3]
```diff
@@ -53,3 +55,81 @@
...
+
+(s/defschema ESConnStateProps
+  (st/optional-keys
+   {:config {s/Any s/Any}
+    :props {s/Keyword s/Any}
+    :index s/Any
+    :conn s/Any
+    :services s/Any}))
+
+(s/defn rename-search-fields :- (s/maybe {s/Keyword [s/Str]})
```
[the remark this answers is not in the document]
Comment (agzam) [reply src/ctia/stores/es/query.clj:106]: Roger that. Planning to add some test(s)

File: src/ctia/stores/es/query.clj (hunk 2 above)
Comment (agzam) [reply src/ctia/stores/es/query.clj:87]: @DeLaGuardo @ereteog I've added tests: test/ctia/stores/es/query_test.clj

Comment (DeLaGuardo) [reply src/ctia/stores/es/query.clj:87]: It is a bit hard to follow the logic in searchable-fields-map-impl. So far I wrap it in my head as "Walk through data structure until value is a map without `:type "keyword"` collecting on the way all the keys except :properties as a path. Then if the leaf node has at least one field of type text - keep the pair, path and field. Later collection of pair will be turned into hash-map." 
Based on that I wrote alternative implementation:

```clj
(defn- searchable-fields-map-impl
  ([m] (searchable-fields-map-impl [] m))
  ([path m]
   (if (and (map? m) (not= (:type m) "keyword"))
     (let [step-fn (fn [k]
                     (let [path (if (= k :properties) path (conj path k))]
                       (searchable-fields-map-impl path (m k))))]
       (mapcat step-fn (keys m)))
     (let [fields (reduce-kv
                   (fn [acc field {:keys [type]}]
                     (if (= type "text")
                       (conj acc field)
                       acc))
                   []
                   (:fields m))]
       (map #(conj [path] %) fields)))))

(s/defn searchable-fields-map :- {s/Str s/Str}
  "Walks through entity's mapping properties and finds all fields with secondary searchable token.
Returns a map where key is path to a field, and value - path to the nested text token."
  [properties :- (s/maybe {s/Keyword s/Any})]
  (into {}
        (map (fn [[path field]]
               (let [path (str/join "." (map name path))]
                 [path
                  (str/join "." [path (name field)])])))
        (searchable-fields-map-impl properties)))
```

@agzam What do you think?

**Comment by ereteog** [comment]
> Do not forget to define searchable-token mappings
> 
> > We need to change the mappings of token default fields to searchable-token and rename them all. There is at least id and observable.value in the common mapping definition.

File: src/ctia/stores/es/query.clj (hunk 2 above)
Comment (agzam) [reply src/ctia/stores/es/query.clj:82]: Very nice @DeLaGuardo, your version looks a lot nicer. Thank you for spending time to figure it out. Switched to your implementation.

**Review by rplevy — APPROVED** [review]
LGTM!

**Comment by agzam** [comment]
> We need to change the mappings of token default fields to searchable-token and rename them all. 

Wait, wait. Are you saying that we need to change `https://github.com/threatgrid/ctia/blob/rewrite-map_for_searchable-tokens/src/ctia/stores/es/mapping.clj#L58`, replacing:
 
```
(def base-entity-mapping
  {:id token
```
with:
```
(def base-entity-mapping
  {:id searchable-token
```
Is that right, @ereteog? Whoa, that's scary.

**Comment by ereteog** [comment]
> Is that right, @ereteog? Whoa, that's scary.

Yes and other tokens fields that will be used as default search field. And yes, that's scary, [that's why iroh services does it :-) ](https://www.youtube.com/watch?v=wMEq1mGpP5A)

**Comment by agzam** [comment]
@ereteog I've checked `searchable-fields` per entity and re-mapped every searchable field to use searchable-token [here](https://github.com/threatgrid/ctia/pull/1167/commits/1da4390f1a237e1c5452fc53e893bb5bbd8ff584)

File: src/ctia/entity/incident.clj [hunk 4]
```diff
@@ -130,15 +130,15 @@
...
-     {:confidence em/token
-      :status em/token
-      :incident_time em/incident-time
-      :categories em/token
-      :discovery_method em/token
-      :intended_effect em/token
-      :assignees em/token
-      :promotion_method em/token
-      :severity em/token})}})
+     {:confidence       em/searchable-token
```
Comment (ereteog) [inline src/ctia/entity/incident.clj:133]:
```suggestion
     {:confidence      searchable-token
```

**Review by ereteog — COMMENTED** [review]
I detected 2 changes on searchable-fields that I detected with the mapping.
In order to gain confidence on this we need a full text search on nested fields. Sightings are a good case, add a full text search on a value of `targets.observables.value`of a sighting maximal.

File: src/ctia/stores/es/query.clj [hunk 5]
```diff
@@ -53,3 +55,76 @@
...
+        term->es-query-part (fn [{:keys [query_mode fields] :as text-query}]
+                              (hash-map
+                               (or query_mode :query_string)
+                               (-> text-query
+                                   (dissoc :query_mode)
+                                   (merge
+                                    (when (and default_operator
+                                               (not= query_mode :multi_match))
+                                      {:default_operator default_operator})
+                                    (rename-search-fields es-conn-state fields)))))]
```
Comment (ereteog) [inline src/ctia/stores/es/query.clj:129]: We need to add the condition of the feature flag and turn it on after the refresh of the mappings.

Comment (agzam) [reply src/ctia/stores/es/query.clj:129]: Same flag or should I create new one 🤔 https://github.com/threatgrid/ctia/blob/master/src/ctia/http/routes/crud.clj#L336

Comment (agzam) [reply src/ctia/stores/es/query.clj]: New flag for now https://github.com/threatgrid/ctia/blob/rewrite-map_for_searchable-tokens/src/ctia/stores/es/query.clj#L131

**Review by rplevy — APPROVED** [review]
LGTM!

File: test/ctia/http/generative/fulltext_search_test.clj [hunk 6]
```diff
@@ -140,13 +140,13 @@
                              (utils/update-items
                               incidents
                               ;; update only the first, leave the rest unchanged
-                              #(assoc % :discovery_method "Log Review"
+                              #(assoc % :short_description "Log Review"
```
Comment (ereteog) [inline test/ctia/http/generative/fulltext_search_test.clj:143]: why did you change these tests?
I prefer the version with mixed tokens and text.

Comment (ereteog) [inline test/ctia/http/generative/fulltext_search_test.clj:143]: please preserve original version, the changes you made should not impact this.

**Review by ereteog — APPROVED** [review]
LGTM

File: test/ctia/http/generative/fulltext_search_test.clj (hunk 6 above)
Comment (agzam) [reply test/ctia/http/generative/fulltext_search_test.clj:143]: > changes you made should not impact this.

They do though, remember, we've limited Incident searchable_fields, `discovery_method` and `assignees` are no longer in `ctia.entity.incident/searchable-fields`, I had to change it.

Comment (ereteog) [reply test/ctia/http/generative/fulltext_search_test.clj:143]: then it would be nice to replace a token field by another token field

Comment (agzam) [reply test/ctia/http/generative/fulltext_search_test.clj:143]: Eh? I'm not sure what you mean, would you elaborate?
